### PR TITLE
0.3.0

### PR DESCRIPTION
**Pull request by BenFradet**
- [x] short-circuit on missing templating variable (#16)
- [x] short-circuit on missing env variable (#26)
- [x] `timeWithFormat` template function (#18)
- [x] base64 template functions (#19)
  - [x] `base64`
  - [x] `base64File`
- [x] tags (#27)
- [x] file-based locking (#20)
- [x] consul-based locking (#17)
- [x] wiki:
  - [x] timeWithFormat
  - [x] base64
  - [x] base64File
  - [x] lock
  - [x] consul
- [x] replace raw github link by iglucentral (depends on snowplow/iglu-central#579)
- [x] update `up.playbooks` with consul (depends on snowplow/ansible-playbooks#123)
- [ ] push the wiki updates

### CODE REVIEW

**Comment by BenFradet** [comment]
@alexanderdean Ready for a review

Otherwise, it's ready to go given:
- https://github.com/snowplow/ansible-playbooks/pull/123
- https://github.com/snowplow/iglu-central/pull/579

File: CHANGELOG [hunk 1]
```diff
@@ -1,3 +1,14 @@
+Version 0.3.0 (2017-05-24)
+--------------------------
+Short-circuit execution on unset variable (#16)
+Short-circuit on unset env variable (#26)
```
Comment (alexanderdean) [inline CHANGELOG]: Should be "Short-circuit execution on ..." for consistency

File: Makefile [hunk 2]
```diff
@@ -13,7 +13,7 @@ build_log     = $(build_dir)/.build
 merge_log     = $(build_dir)/.merge
 
 cluster_avsc_link  = "http://iglucentral.com/schemas/com.snowplowanalytics.dataflowrunner/ClusterConfig/avro/1-1-0"
-playbook_avsc_link = "http://iglucentral.com/schemas/com.snowplowanalytics.dataflowrunner/PlaybookConfig/avro/1-0-0"
+playbook_avsc_link = "https://raw.githubusercontent.com/snowplow/iglu-central/feature/df-runner-0.3/schemas/com.snowplowanalytics.dataflowrunner/PlaybookConfig/avro/1-0-1"
```
Comment (alexanderdean) [inline Makefile]: I'll do the Iglu Central release later today

File: src/config_resolver.go [hunk 3]
```diff
@@ -33,8 +37,25 @@ var (
...
+		},
+		"base64": func(src string) string {
+			return base64.StdEncoding.EncodeToString([]byte(src))
+		},
+		"base64File": func(filename string) (string, error) {
+			content, err := ioutil.ReadFile(filename)
+			if err != nil {
+				return "", err
+			}
+			return base64.StdEncoding.EncodeToString(content), nil
```
Comment (alexanderdean) [inline src/config_resolver.go]: Nice

File: src/main.go [hunk 4]
```diff
@@ -244,19 +268,33 @@ func up(emrConfig string, vars string) (string, error) {
...
 
+	if !async && lock != "" {
+		l, err := getLock(consul, lock)
+		err = l.TryLock()
+		if err != nil {
+			return err
+		}
+		defer l.Unlock()
+	} else if lock != "" {
+		return errors.New("--" + fAsync + " and --" + fLock + " are not compatible")
```
Comment (alexanderdean) [inline src/main.go]: Fair play.

File: src/main.go [hunk 5]
```diff
@@ -211,6 +221,20 @@ func getAsyncFlag() cli.BoolFlag {
 	return cli.BoolFlag{Name: fAsync, Usage: "Asynchronous execution of the jobflow steps"}
 }
 
+func getLockFlag() cli.StringFlag {
+	return cli.StringFlag{
+		Name:  fLock,
+		Usage: "Name of the lock held for the duration of the jobflow steps",
```
Comment (alexanderdean) [inline src/main.go]: "of the jobflow steps. Stored in a file or in Consul depending on the --" + fConsul + " flag"

File: src/lock.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// FileLock is for file-based locks
+type FileLock struct {
+	lock lockfile.Lockfile
+}
+
+// InitFileLock builds a FileLock (a temp file) with the argument as name
+func InitFileLock(name string) (Lock, error) {
+	lock, err := lockfile.New(filepath.Join(os.TempDir(), name))
```
Comment (alexanderdean) [inline src/lock.go]: I don't think a temp dir is a safe place for a lock. Rather we should assume that the supplied lock path is an absolute path.

Comment (BenFradet) [inline src/lock.go]: So we wouldn't allow relative paths?

Comment (alexanderdean) [inline src/lock.go]: Hmm, actually we should probably allow relative paths...

**Comment by BenFradet** [comment]
@alexanderdean I think I've taken care of everything

**Comment by alexanderdean** [comment]
LGTM! Can we get an -rc1 of this into Bintray please?

**Comment by BenFradet** [comment]
rc1 is on bintray

**Comment by alexanderdean** [comment]
Please:

* Rename existing lock behavior to `-softLock`
* Add additional lock behavior where lock is **not** deleted if job fails; this should be accessed as `-lock` argument

**Comment by BenFradet** [comment]
rc2 released

**Comment by BenFradet** [comment]
rc3 should land in bintray soon :tm:

**Comment by alexanderdean** [comment]
Hey @BenFradet - apologies for the delay in thinking of this requirement, but we need to come up with a dedicated return code (e.g. 3 or 4) for the scenario that the job cannot run because there is an existing lock in place. It might also make sense to make any informational message in this scenario a WARN rather than an ERROR.

Why? This is so that a job runner tool like Factotum can distinguish between a hard failure (e.g. EMR job failed) and a no-op situation (job could not start because existing job is running). That distinction is super-important because we cannot afford to raise a pager duty alert for every single no-op - to put it another way, if my EMR job takes 15 minutes and I schedule the job every 5 minutes, then I *expect* 8 no-ops an hour.

Sorry for the delayed requirement!

**Comment by BenFradet** [comment]
Should this be a new issue since the exit code is always 0 atm?

**Comment by alexanderdean** [comment]
Hmm - I think this is covered by the lock ticket - as it's just a property of the lock handling?

**Comment by BenFradet** [comment]
don't you need exit code = 1 for a true failure?

**Comment by alexanderdean** [comment]
Oh lord, sorry I just saw what you meant. So Dataflow Runner can never return non-zero?! :godmode:

**Comment by BenFradet** [comment]
not according to https://github.com/urfave/cli#exit-code

**Comment by alexanderdean** [comment]
Ouch - new ticket then please, this is a bit of a showstopper!

**Comment by BenFradet** [comment]
my bad the logging library (!!) applies exit code https://github.com/sirupsen/logrus#level-logging

**Comment by alexanderdean** [comment]
Hmm interesting, can we still do the exit codes in the other ticket though? I don't think we necessarily need an independent ticket though.

**Comment by BenFradet** [comment]
As your request has shown this isn't viable in the long term (delegating exit code to the logging library).
As a result, I think this ticket is warranted because I'm currently moving away from the above.

**Comment by alexanderdean** [comment]
Cool thanks

**Comment by BenFradet** [comment]
@alexanderdean published rc4 :+1:

**Comment by alexanderdean** [comment]
More edgecases - `timeWithFormat` should be able to take `epoch` as a string. This is because all of the incoming variables provided by `--vars` are stringly typed.

There could still be a usecase for supporting an `int64` as well, but really we could live without it (because we can always put "" around a literal value).

**Comment by alexanderdean** [comment]
Me again! I think I am seeing an exit code of 17 if there is an error creating the lock file (e.g. because the directory of the lock does not exist). A quick look at the code suggests that Dataflow Runner exits 17 if there is any problem with creating the lock, not just the lock file is already there?

**Comment by BenFradet** [comment]
Here's what I currently have with rc5:

- first run is without consul
- 2nd run is with consul with the steps failing
- 3rd run is with consul with the lock created during the 2nd run

df

**Comment by alexanderdean** [comment]
Thanks @BenFradet, I will give this one a spin. What does "Lock already hel" mean - why is this truncated?

**Comment by BenFradet** [comment]
It was a typo during the tests which is fixed in the published artifact, it should just say `Lock already held`.

**Comment by alexanderdean** [comment]
Cool thanks

**Pull request merged**